SeQuant/core/space.hpp

```diff
@@ -417,109 +314,446 @@ class IndexSpace {
+    return base_key_;
+  }
+  static std::wstring_view reduce_key(std::wstring_view key) {
+    const auto underscore_position = key.rfind(L'_');
+    if(underscore_position != -1){ // key can be reduced
+      return key.substr(0, underscore_position);
+    }
+    else{
+      return key;
+    }
+  }
+
+  // having approximate sizes of spaces can be helpful when optimizing evaluation order.
+  unsigned long get_approximate_size() const {
+    return approximate_size_;
   }
 
  private:
   Attr attr_ = Attr::invalid();
+  std::wstring base_key_;
+  unsigned long approximate_size_;
   /// @brief constructs an instance of an IndexSpace object
-  explicit IndexSpace(Attr attr) noexcept : attr_(attr) {
-    assert(attr.is_valid());
+  //explicit IndexSpace(Attr attr) noexcept : attr_(attr) {
+  //  assert(attr.is_valid());
+  //}
+
+
+
+  static std::wstring to_wstring(std::wstring_view key) {
+    return std::wstring(key.begin(), key.end());
   }
 
-  static container::map<Attr, std::wstring> attr2basekey_;
-  static container::map<std::wstring, Attr, KeyCompare> key2attr_;
-  static container::map<Attr, IndexSpace> instances_;
-  static IndexSpace null_instance_;
+};
 
-  static std::wstring_view reduce_key(std::wstring_view key) {
-    const auto underscore_position = key.rfind(L'_');
-    return key.substr(0, underscore_position);
+// An IndexSpaceRegistry is a context specific, modifiable, list of registered indices.
+// It is the task of the USER to ensure that all spaces for their context are included.
+// the result of certain set operations such as intersection and unIon require access to the registry
+//
+class IndexSpaceRegistry{
+ public:
+  IndexSpaceRegistry(){
+    //register nulltype
+    this->add(nulltype);
   }
 
-  /// @param key the key associated with a registered IndexSpace; this can be
-  /// either
-  ///            the base key used to invoke `IndexSpace::register_instance()`
-  ///            or a key used to invoke `IndexSpace::register_key()`
-  /// @return the attribute of the IndexSpace object corresponding to @p key
-  /// @throw bad_key if \p key has not been registered
-  static Attr to_attr(std::wstring_view key) {
-    const auto found_it = key2attr_.find(key);
-    if (found_it != key2attr_.end()) return found_it->second;
-    throw bad_key();
+  //retrieve an IndexSpace from the registry by the label
+  IndexSpace retrieve(std::wstring_view label) const{
+    for(auto it = label_space.begin(); it != label_space.end(); ++it){
+      if(IndexSpace::reduce_key(label) == it->first){
+        return it->second;
+      }
+    }
+    throw std::invalid_argument("The label you are trying to retrieve has not been registered");
+    return label_space.at(L"");
   }
 
-  static std::wstring to_wstring(std::wstring_view key) {
-    return std::wstring(key.begin(), key.end());
+  //add an IndexSpace to this registry. duplicate type bitsets forbidden
+  void add(IndexSpace IS){
+    if(!(find_IndexSpace(IS) == nulltype) ){
```

Comment:
```suggestion
    if(find_IndexSpace(IS) != nulltype){
```
(for easier readability - the leading ! can be easily missed along the parenthesis)